File: src/tools/web_images/ranking.rs
fn rank_candidates(query: &str, candidates: &mut [ImageCandidate]) {
    candidates.sort_by(|left, right| {
        score_candidate(query, right)
            .partial_cmp(&score_candidate(query, left))
            .unwrap_or(std::cmp::Ordering::Equal)
    });
}

fn score_candidate(query: &str, candidate: &ImageCandidate) -> f32 {
    let metadata = format!(
        "{} {} {}",
        candidate.title, candidate.page_url, candidate.image_url
    )
    .to_ascii_lowercase();
    let mut score = 0.0;
    for term in image_query_terms(query) {
        if candidate.title.to_ascii_lowercase().contains(&term) {
            score += 24.0;
        } else if metadata.contains(&term) {
            score += 10.0;
        }
    }
    let short = candidate.width.min(candidate.height);
    let area = candidate.width.saturating_mul(candidate.height);
    score += if short >= 900 {
        28.0
    } else if short >= 600 {
        24.0
    } else if short >= 300 {
        18.0
    } else if short >= 100 {
        4.0
    } else {
        -8.0
    };
    if area >= 1_000_000 {
        score += 7.0;
    }
    let noisy = [
        "thumb",
        "thumbnail",
        "sprite",
        "placeholder",
        "banner",
        "advert",
        "favicon",
    ];
    if noisy.iter().any(|term| metadata.contains(term)) {
        score -= 8.0;
    }
    if metadata.contains("avatar")
        && !query.contains("头像")
        && !query.to_ascii_lowercase().contains("avatar")
    {
        score -= 8.0;
    }
    score
}

fn image_query_terms(query: &str) -> Vec<String> {
    let generic = [
        "图片",
        "照片",
        "高清",
        "壁纸",
        "photo",
        "image",
        "images",
        "picture",
        "wallpaper",
        "hd",
        "4k",
    ];
    query
        .split(|ch: char| ch.is_whitespace() || ch.is_ascii_punctuation())
        .map(|term| term.trim().to_ascii_lowercase())
        .filter(|term| term.len() >= 2 && !generic.contains(&term.as_str()))
        .collect()
}

fn dedupe_candidates(candidates: Vec<ImageCandidate>) -> Vec<ImageCandidate> {
    let mut seen = HashSet::new();
    let mut deduped = Vec::new();
    for candidate in candidates {
        let key = candidate
            .image_url
            .split('?')
            .next()
            .unwrap_or(&candidate.image_url)
            .to_ascii_lowercase();
        if seen.insert(key) {
            deduped.push(candidate);
        }
    }
    deduped
}

fn image_candidate_pool_limit(count: usize) -> usize {
    count.max((count * 4).max(count + 8).min(30))
}

fn image_download_probe_limit(count: usize) -> usize {
    count.max((count * 4).max(count + 6).min(16))
}

fn candidate_json(candidate: ImageCandidate) -> Value {
    json!({
        "title": candidate.title,
        "page_url": candidate.page_url,
        "image_url": candidate.image_url,
        "thumbnail_url": candidate.thumbnail_url,
        "source": candidate.source,
        "width": candidate.width,
        "height": candidate.height,
        "search_description": candidate.search_description,
    })
}

fn stored_json(item: StoredImage) -> Value {
    json!({
        "title": item.candidate.title,
        "page_url": item.candidate.page_url,
        "image_url": item.candidate.image_url,
        "thumbnail_url": item.candidate.thumbnail_url,
        "source": item.candidate.source,
        "local_path": item.local_path,
        "mime_type": item.mime_type,
        "width": item.candidate.width,
        "height": item.candidate.height,
        "size_bytes": item.size_bytes,
        "size_human": format_bytes(item.size_bytes),
        "sha256": item.sha256,
        "used_thumbnail": item.used_thumbnail,
        "search_description": item.candidate.search_description,
        "vision": {
            "status": item.vision.status,
            "accepted": item.vision.accepted,
            "description": item.vision.description,
            "reason": item.vision.reason,
            "provider_id": item.vision.provider_id,
            "model": item.vision.model,
            "error": item.vision.error,
        },
    })
}

File: plugins/web-images/ranking.lua
local M = {}
local generic = { ["图片"]=true, ["照片"]=true, ["高清"]=true, ["壁纸"]=true,
    photo=true, image=true, images=true, picture=true, wallpaper=true, hd=true, ["4k"]=true }

--- 【网页搜图】【查询词】按 Unicode 空白与 ASCII 标点分词，去掉原通用搜图词
--- @param query string 用户查询
--- @return table 保持原顺序的查询词
function M.terms(query)
    local parts = {}
    for index = 1, #query do
        local byte = query:byte(index)
        local punctuation = (byte >= 33 and byte <= 47) or (byte >= 58 and byte <= 64)
            or (byte >= 91 and byte <= 96) or (byte >= 123 and byte <= 126)
        parts[index] = punctuation and " " or query:sub(index, index)
    end
    local result = sai.json.array()
    for term in sai.text.collapse_whitespace(table.concat(parts)):lower():gmatch("[^ ]+") do
        if #term >= 2 and not generic[term] then result[#result + 1] = term end
    end
    return result
end

--- 【网页搜图】【候选评分】保持标题命中、尺寸、噪声与头像规则
--- @param query string 用户查询
--- @param item table 图片候选
--- @return number 原权重得分
function M.score(query, item)
    local title = item.title:lower()
    local metadata = (item.title .. " " .. item.page_url .. " " .. item.image_url):lower()
    local score = 0
    for _, term in ipairs(M.terms(query)) do
        if title:find(term, 1, true) then score = score + 24
        elseif metadata:find(term, 1, true) then score = score + 10 end
    end
    local short = math.min(item.width, item.height)
    score = score + (short >= 900 and 28 or short >= 600 and 24 or short >= 300 and 18 or short >= 100 and 4 or -8)
    if item.width * (item.height + 0.0) >= 1000000 then score = score + 7 end
    for _, term in ipairs({"thumb", "thumbnail", "sprite", "placeholder", "banner", "advert", "favicon"}) do
        if metadata:find(term, 1, true) then score = score - 8; break end
    end
    if metadata:find("avatar", 1, true) and not query:find("头像", 1, true) and not query:lower():find("avatar", 1, true) then
        score = score - 8
    end
    return score
end

--- 【网页搜图】【地址去重】忽略查询参数和 ASCII 大小写，保留首个候选
--- @param candidates table 合并后的候选数组
--- @return table 去重后的候选数组
function M.dedupe(candidates)
    local result, seen = sai.json.array(), {}
    for _, item in ipairs(candidates) do
        local key = item.image_url:match("^[^?]*"):lower()
        if not seen[key] then result[#result + 1], seen[key] = item, true end
    end
    return result
end

--- 【网页搜图】【稳定排序】分数相同时保持搜索引擎原顺序
--- @param query string 用户查询
--- @param candidates table 待排序数组
--- @return table 新的有序候选数组
function M.rank(query, candidates)
    local ranked = {}
    for index, item in ipairs(candidates) do ranked[index] = {item=item, score=M.score(query, item), index=index} end
    table.sort(ranked, function(left, right)
        return left.score > right.score or (left.score == right.score and left.index < right.index)
    end)
    local result = sai.json.array()
    for _, entry in ipairs(ranked) do result[#result + 1] = entry.item end
    return result
end

--- 【网页搜图】【候选预算】按用户数量扩大候选池，最多三十条
--- @param count integer 最终请求数量
--- @return integer 候选池上限
function M.pool_limit(count)
    return math.max(count, math.min(math.max(count * 4, count + 8), 30))
end

--- 【网页搜图】【下载预算】限制尝试数量，同时保留小请求的替代候选
--- @param count integer 最终请求数量
--- @return integer 下载尝试上限
function M.probe_limit(count)
    return math.max(count, math.min(math.max(count * 4, count + 6), 16))
end

return M
